dartagnan/src/main/java/com/dat3m/dartagnan/parsers/cat/VisitorBase.java

```diff
@@ -0,0 +1,358 @@
+            Axiom axiom = (Axiom) constructor.newInstance(r, ctx.negate != null, ctx.flag != null);
+            if (ctx.NAME() != null) {
+                axiom.setName(ctx.NAME().toString());
+            }
+            wmm.addAxiom(axiom);
+        } catch (NoSuchMethodException | InstantiationException | IllegalAccessException |
+                 InvocationTargetException e) {
+            throw new ParsingException(ctx.getText());
+        }
+        return null;
+    }
+
+    @Override
+    public Void visitLetDefinition(LetDefinitionContext ctx) {
+        String n = ctx.n.getText();
+        alias = ctx.n.getText();
+        Object o = ctx.e.accept(this);
+        namespace.put(n, o);
+        if (o instanceof FilterAbstract) {
+            FilterAbstract f = (FilterAbstract) o;
+            f.setName(n);
+            wmm.addFilter(f);
+        }
+        return null;
+    }
+
+    @Override
+    public Void visitLetRecDefinition(LetRecDefinitionContext ctx) {
+        int size = ctx.letRecAndDefinition().size();
+        Relation[] recursiveGroup = new Relation[size + 1];
+        String n = ctx.n.getText();
+        recursiveGroup[0] = wmm.newRelation(n);
+        namespace.put(n, recursiveGroup[0]);
+        for (int i = 0; i < size; i++) {
+            n = ctx.letRecAndDefinition(i).n.getText();
+            recursiveGroup[i + 1] = wmm.newRelation(n);
+            namespace.put(n, recursiveGroup[i + 1]);
+        }
+        for (Relation r : recursiveGroup) {
+            r.setRecursive();
+        }
+        current = recursiveGroup[0];
+        relation(ctx.e);
+        for (int i = 0; i < size; i++) {
+            current = recursiveGroup[i + 1];
+            relation(ctx.letRecAndDefinition(i).e);
+        }
+        return null;
+    }
+
+    @Override
+    public Object visitExpr(ExprContext ctx) {
+        return ctx.e.accept(this);
+    }
+
+    @Override
+    public Object visitExprBasic(ExprBasicContext ctx) {
+        String n = ctx.n.getText();
+        Object o = namespace.get(n);
+        if (o != null) {
+            addCurrentName(o);
+            return o;
+        }
+        Relation r = wmm.getRelation(n);
+        if (r != null) {
+            namespace.put(n, r);
+            addCurrentName(r);
+            return r;
+        }
+        FilterBasic f = FilterBasic.get(n);
+        namespace.put(n, f);
+        addCurrentName(f);
+        return f;
+    }
+
+    @Override
+    public Object visitExprIntersection(ExprIntersectionContext c) {
+        checkNoCurrentOrNoAlias(c);
+        Relation r = current();
+        String a = alias();
+        Object o1 = c.e1.accept(this);
+        Object o2 = c.e2.accept(this);
+        if (o1 instanceof Relation) {
+            Relation r0 = r != null ? r : newRelation(a);
+            return addDefinition(new Intersection(r0, (Relation) o1, relation(o2, c)));
+        }
+        return withAlias(a, FilterIntersection.get(set(o1, c), set(o2, c)));
+    }
+
```

Comment:
Partially.  `checkNoCurrentOrNoAlias` checks for the structure invariant of the visitor.  `let rec` contexts pre-create all recursive relations and pass the respective to-be-defined one to the expression context.  `let` contexts pass just the desired name to avoid the creation of a new relation, in case aliasing will take place.  Both cannot be active at the same time.
Maybe it should call `verify`, instead of throwing `ParsingException`.

Reply:
It now throws `VerifyException`, to emphasize that this is the programmer's fault.